React frontend changes

GreatDanton opened the pull request:
FFT filters are now applied to url
Project device fields are now editable within the table

tangkong commented:
Jan, let us know when you're finished pushing to this branch and we'll try to get a review through.  (since it looks like you're still working on this)

GreatDanton commented:
> Jan, let us know when you're finished pushing to this branch and we'll try to get a review through. (since it looks like you're still working on this)

You can check right away, since I am done for now. I'll start working on diff page next.

laura-king commented on react/src/app/projects/[projectId]/project_details.tsx:
```diff
@@ -288,324 +410,330 @@ export const ProjectSpecificPage: React.FC<{ projectId: string }> = ({ projectId
-            submitSearchForm();
+const DeviceDataTableRow: React.FC<{ project?: ProjectInfo, device: ProjectDeviceDetails, disabled: boolean, onEdit: (device: ProjectDeviceDetails) => void, onCopyFft: (device: ProjectDeviceDetails) => void }> = ({ project, device, disabled, onEdit, onCopyFft }) => {
+    // we have to cache each table row, as once we have lots of rows in a table editing text fields within
+    // becomes very slow due to constant rerendering of rows and their tooltips on every keystroke. 
```
Just curious, is there any other way around this? Will this continue to get slower as a project grows?

GreatDanton replied:
Caching allows us to avoid the slowdowns that appeared during the row editing. We shouldn't have any more performance problems with this table.

laura-king commented on react/src/app/projects/[projectId]/project_details.tsx:
```diff
@@ -1,17 +1,79 @@
-import { useEffect, useMemo, useState } from "react";
-import { ProjectApprovalDialog } from "../project_approval_dialog";
-import { FFT, FFTDiff, ProjectDeviceDetails, ProjectInfo, fetchAllProjects, fetchProjectDiff, isProjectSubmitted, parseFftFieldNameFromFftDiff, parseFftIdFromFftDiff } from "../project_model";
+import { Alert, Button, ButtonGroup, Colors, Divider, HTMLSelect, Icon, InputGroup, NonIdealState, NumericInput } from "@blueprintjs/core";
```
Is Blueprintjs the library standard to interact with dialogs/modals/inputs in react?

GreatDanton replied:
There is no standard library, it's just one of the possible component libraries to use for that. I just think blueprint is nicer to use over others I've seen.

laura-king commented on react/src/app/projects/[projectId]/project_details.tsx:
```diff
@@ -288,324 +410,330 @@ export const ProjectSpecificPage: React.FC<{ projectId: string }> = ({ projectId
+        setSubmitting(true);
+        syncDeviceUserChanges(project._id, deviceWithChanges.id, changes)
+            .then((response) => {
+                // TODO: for some reason server returns data for all devices again
```
I know exactly where this happens. It's in the endpoint svc_update_fc_in_project() in licco.py/services. 

If it causes slowdowns, it's a very easy fix to either edit this endpoint or make another one that returns anything that the frontend needs

GreatDanton replied:
For now I didn't notice this endpoint being slow; if we change it now we also break the old gui but long term I guess we can change it.

laura-king commented:
It looks like the FFT sync feature within the project details page only lets you sync with projects of the "submitted" status

laura-king commented:
The ray trace field allows me to save a decimal. The field should only allow 0 and non-negative integers. 

Not sure where the control for this was in the previous version

laura-king commented:
This is a small nitpick- I think that on the project details page, the project name blends in with the other headers and is difficult to distinguish. I think it would be helpful if it was either larger, another color or font, not sure.

laura-king commented:
I'm only able to sort the X/Y/Z values from highest -> lowest. When I click the sort again, the table doesn't change. The other columns seem fine

laura-king commented:
I am unable to find a way to add a new FFT/device to a project from within the project details page. Is there a way to do this in the new UI?

In the old client, this functionality is within the button/modal represented by the magnifying glass icon with the plus inside of it.

laura-king commented:
For the record, all of my testing is on Firefox. Not sure if any of these are browser-specific issues

GreatDanton commented:
> It looks like the FFT sync feature within the project details page only lets you sync with projects of the "submitted" status. Ideally, this would have all projects in development, submitted, and approved status.

Done


> The ray trace field allows me to save a decimal. The field should only allow 0 and non-negative integers.

Fixed

> This is a small nitpick- I think that on the project details page, the project name blends in with the other headers and is difficult to distinguish. I think it would be helpful if it was either larger, another color or font, not sure.

Changed


> I'm only able to sort the X/Y/Z values from highest -> lowest. When I click the sort again, the table doesn't change. The other columns seem fine

Fixed, it was a bug in the column type detection.


> I am unable to find a way to add a new FFT/device to a project from within the project details page. Is there a way to do this in the new UI?
>
> In the old client, this functionality is within the button/modal represented by the magnifying glass icon with the plus inside of it.

Dialog was added with the latest commit

tangkong commented:
> The ray trace field allows me to save a decimal. The field should only allow 0 and non-negative integers.

Perhaps a dumb question: why was this not originally just a boolean input?  or a checkbox?

laura-king commented:
> > The ray trace field allows me to save a decimal. The field should only allow 0 and non-negative integers.
> 
> Perhaps a dumb question: why was this not originally just a boolean input? or a checkbox?

Originally it actually was a boolean. But stakeholders decided that it must specifically be a non-negative integer, or a 0-I remember Kai requesting this. Suyun may be able to answer.

laura-king commented:
Just looked at all the fixes, thanks for taking care of them all so quick!

Another bug I found- When a device is in any state other than "Conceptual", certain fields are required to have valid data in them. This is currently checked correctly when the device switches from Conceptual -> ReadyForInstallation. 

However, once a device is already in a Non-Conceptual state, the system appears to allow a user to remove required data. 

Steps:
1. Switch a device to ReadyForInstallation (with all required data)
2. Edit the device and remove a required field of data (nominal angle z, for example)
3. Ideally we get an error popup or block, but it currently appears to the end user that the data is changed. 

It looks like it doesn't actually edit/remove the data from the database, so we probably just need a pop-up/error handler here

laura-king commented:
We also need the add FFT button/dialog on each project details page as well. It would add a currently existing FFT in to the project, or if it does not currently exist in the DB, would make a new FFT in the database, and add in a blank version to the project with a default state value of Conceptual. 

This button should be working in the old version if my description is confusing

laura-king commented:
Also, there is some features needed around the approved project in the projects overview page. Sorry for not having all of these things defined ahead of time.

1. We need some way to distinguish the currently approved project from the projects page. In the previous version the row was highlighted orange, but we're not tied to that color/implementation. Just need some way to visually distinguish the approved project.  A nice to have(it's a requested feature) would be to have it pinned to the top of the page at all times.
2. The Approved Project should have the same buttons disabled as Submitted projects. (Only Clone, Download, and Diff buttons enabled)

If you can't get the approval process working in the old version to set a project to that state, you can just change the status from the mongo shell via command line

laura-king commented:
The "Create a New Project" dialog also should allow two options as a base: a Blank Project, or the Approved Project.

GreatDanton commented:
> Another bug I found- When a device is in any state other than "Conceptual", certain fields are required to have valid data in them. This is currently checked correctly when the device switches from Conceptual -> ReadyForInstallation.

Fixed, undefined values (removed fields) were not serialized into JSON correctly.

> We need some way to distinguish the currently approved project from the projects page. 

Done

> The "Create a New Project" dialog also should allow two options as a template: a Blank Project, or the Approved Project.

Done
